lib/structs/events/collections.cpp

```diff
@@ -330,9 +331,15 @@ from_json(const nlohmann::json &obj, TimelineEvent &e)
                 break;
             }
             case MsgType::Unknown: {
+                e.data = events::RoomEvent<events::msg::Unknown>(obj);
+                break;
+            }
+            case MsgType::Redacted: {
                 e.data = events::RoomEvent<events::Unknown>(obj);
                 break;
             }
+            case MsgType::Invalid:
+                break;
```

Comment:
So in general it looks good, but since these are where the interesting stuff happens, these should really be covered by tests (which should also make codecov pass).

Reply:
OK, I will try to figure out a proper test for that.

Reply:
Actually, now that I see the function, I see that this would leave the .data member uninitialized. I thought we were looking at the `parse_timeline_events` function, which adds them to a list. I think this specific function always needs to return something, even if that is `events::RoomEvent<events::Unknown>(obj)`

Reply:
> I think this specific function always needs to return something, even if that is `events::RoomEvent<events::Unknown>(obj)`

No, if you look a little farther down there is a whole list of event types that don't have any `e.data` set (see lines 382-398).

> But you should probably add a test for both :3

I'm confused. Both what?

Reply:
parse_timeline_events and the direct from_json functions. one is used for /sync, where you want to filter out invalid events and another one is uses for parsing stuff directly from a json event.

Reply:
OK, I have parse_timeline_events working. I can continue to look into from_json, but the only room message case in that function that is covered (and indeed, one of only 5 covered cases in that function) is Text. That's going to be a bit tricky to track down, but I certainly can try if you think I should.